scio-avro/src/main/scala/com/spotify/scio/avro/ProtobufIO.scala

```diff
@@ -53,13 +71,51 @@ final case class ProtobufIO[T <: Message: ClassTag](path: String) extends ScioIO
     data.write(underlying)(params.copy(metadata = metadata)).underlying
   }
 
-  override def tap(read: ReadP): Tap[T] =
-    ProtobufFileTap(path, read)
+  override def tap(read: ReadP): Tap[T] = ProtobufFileTap(path, read)
 }
 
-object ProtobufIO {
-  type ReadParam = GenericRecordIO.ReadParam
-  val ReadParam = GenericRecordIO.ReadParam
-  type WriteParam = GenericRecordIO.WriteParam
-  val WriteParam = GenericRecordIO.WriteParam
+final case class ProtobufTypedObjectFileIO[T: Coder, U <: Message: ClassTag](
+  path: String
+)(implicit pt: ProtobufType[T, U])
+    extends ProtobufIO[T] {
+  override type ReadP = ProtobufTypedObjectFileIO.ReadParam
+  override type WriteP = ProtobufTypedObjectFileIO.WriteParam
+  override def testId: String = s"ProtobufIO($path)"
+
+  private lazy val underlying: ObjectFileIO[U] = ObjectFileIO(path)
+
+  /**
+   * Get an SCollection for a Protobuf file.
+   *
+   * Protobuf messages are serialized into `Array[Byte]` and stored in Avro files to leverage Avro's
+   * block file format.
+   */
+  override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+    sc.transform(_.read(underlying)(params).map(u => pt.from(u)))
+
+  /**
+   * Save this SCollection as a Protobuf file.
+   *
+   * Protobuf messages are serialized into `Array[Byte]` and stored in Avro files to leverage Avro's
+   * block file format.
+   */
+  override protected def write(data: SCollection[T], params: WriteP): Tap[T] = {
+    val metadata = params.metadata ++ ProtobufUtil.schemaMetadataOf[U]
+    underlying
+      .writeWithContext(
```

Comment:
shouldn't this just be `.write` ?

Reply:
See https://github.com/spotify/scio/pull/5286#discussion_r1527223665